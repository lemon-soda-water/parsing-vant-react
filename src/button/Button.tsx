import { cloneElement, useContext, useMemo } from 'react'
import { createNamespace } from '../utils/create'
import { ButtonProps } from './PropsType'
import ButtonContext from './ButtonContext'
import clsx from 'clsx'
import { BORDER_SURROUND, SHADOW, WHITE } from '../utils/constant'
import Loading from '../Loading'

const [bem] = createNamespace('button')

const Button: React.FC<ButtonProps> = (props: ButtonProps) => {
  const {
    color,
    loading,
    className,
    hairline,
    loadingText,
    round,
    square,
    plain: _plain,
    block: _block,
    loadingType,
    nativeType: _nativeType,
    iconPosition: _iconPosition,
    ...restProps
  } = props

  const { parent } = useContext(ButtonContext)

  /* 
    useMemo(cb, [dep])

    只有当 dep 的值发生改变的时候，cb才会重新调用
  */
  // 按钮大小
  const size = useMemo(
    () => props.size || parent?.size || 'normal',
    [parent?.size, props.size],
  )

  // 按钮类型
  const type = useMemo(
    () => props.type || parent?.type || 'default',
    [parent?.type, props.type],
  )

  // 是否为朴素元素
  const plain = useMemo(() => _plain ?? parent?.plain, [parent?.plain, _plain])

  // 是否为块级元素
  const block = useMemo(() => _block ?? parent?.block, [parent?.block, _block])

  // 图标的位置
  const iconPosition = useMemo(
    () => _iconPosition || parent?.iconPosition || 'left',
    [parent?.iconPosition, _iconPosition],
  )

  // 是否禁用按钮
  const disabled = useMemo(
    () => props.disabled ?? parent?.disabled,
    [parent?.disabled, props.disabled],
  )

  // 原生 button 标签的 type 属性
  const nativeType = useMemo(
    () => _nativeType || parent?.nativeType || 'button',
    [parent?.nativeType, _nativeType],
  )

  // html元素的标签名，这里默认是 button
  const tag = useMemo(
    () => props.tag || parent?.tag || 'button',
    [parent?.tag, props.tag],
  )

  // 标签名 默认是 button
  const TagElement = tag as React.ElementType

  // 合并 className
  const classes = clsx(
    className,
    bem([
      type,
      size,
      {
        plain,
        loading,
        disabled,
        hairline,
        block,
        round,
        square,
      },
    ]),
    { [BORDER_SURROUND]: hairline },
    props.shadow && `${SHADOW}--${+props.shadow}`,
  )
  // Record 构建一个对象类型，key 是 string 类型， value 是 string 或 number 类型
  const style: Record<string, string | number> = { ...props.style }

  if (color) {
    // plain 判断是否为朴素元素  字体颜色
    style.color = plain ? color : WHITE

    if (!plain) {
      // Use background instead of backgroundColor to make linear-gradient work
      style.background = color
    }

    // 当 color 为 linear-gradient 时隐藏边框
    if (color.indexOf('gradient') !== -1) {
      style.border = 0
    } else {
      style.borderColor = color
    }
  }

  // 点击事件
  const onClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    // 当不是加载状态，不是禁用状态，并且 有传递点击事件
    if (!loading && !disabled && props.onClick) {
      // 调用点击事件，并把事件对象传递回去
      props.onClick(event)
    }
  }

  const renderLoadingIcon = () => {
    if (loading) {
      const { loadingSize = '20px' } = props
      return (
        <Loading
          className={clsx(bem('loading'))}
          size={loadingSize}
          type={loadingType}
          color={type === 'default' ? undefined : ''}
        />
      )
    }
    return null
  }

  const renderIcon = () => {
    if (props.loading) {
      return renderLoadingIcon()
    }

    if (props.icon) {
      return cloneElement(props.icon as React.ReactElement, {
        className: clsx(bem('icon')),
      })
    }

    return null
  }

  // 渲染文本
  const renderText = () => {
    let text
    if (loading) {
      text = loadingText
    } else {
      text = props.children || props.text
    }

    if (text) {
      return (
        <span key="text" className={clsx(bem('text'))}>
          {text}
        </span>
      )
    }
    return null
  }

  const renderContent = () => (
    <div className={clsx(bem('content'))}>
      {iconPosition === 'left' && renderIcon()}
      {renderText()}
      {iconPosition === 'right' && renderIcon()}
    </div>
  )

  return (
    <TagElement
      {...restProps}
      disabled={disabled}
      className={classes}
      style={style}
      type={nativeType}
      onClick={onClick}
    >
      {renderContent()}
    </TagElement>
  )
}

export default Button
