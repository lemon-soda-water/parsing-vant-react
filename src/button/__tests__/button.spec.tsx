import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import Button from '../index'

describe('button test', () => {
  it('Successful rendering of the button component', () => {
    const wrapper = render(<Button>按钮</Button>)

    expect(wrapper.container.innerHTML).toMatchSnapshot()
  })

  it('should trigger a click event', () => {
    const clickFn = jest.fn(() => {})

    const wrapper = render(<Button onClick={clickFn}>按钮</Button>)
    wrapper.getByRole('button').click()

    expect(clickFn).toHaveBeenCalled()
    expect(wrapper.container.innerHTML).toMatchSnapshot()
  })

  it('should not trigger click event when disabled', () => {
    const clickFn = jest.fn(() => {})

    const props = {
      disabled: true,
      onClick: clickFn,
    }

    const wrapper = render(<Button {...props}>按钮</Button>)
    wrapper.getByRole('button').click()

    expect(clickFn).not.toHaveBeenCalled()
    expect(wrapper.container.innerHTML).toMatchSnapshot()
  })

  it('should not trigger click event when loading', () => {
    const clickFn = jest.fn(() => {})

    const props = {
      loading: true,
      onClick: clickFn,
    }

    const wrapper = render(<Button {...props}>按钮</Button>)
    wrapper.getByRole('button').click()

    expect(clickFn).not.toHaveBeenCalled()
    expect(wrapper.container.innerHTML).toMatchSnapshot()
  })

  it('should hide border when color is gradient', () => {
    const props = {
      color: 'linear-gradient(#000, #fff)',
    }

    const wrapper = render(<Button {...props}>按钮</Button>)
    const button = wrapper.getByRole('button')

    expect(button.style.border).toEqual('0px')
    expect(wrapper.container.innerHTML).toMatchSnapshot()
  })

  it('should render loading slot correctly', () => {
    const props = {
      loading: true,
      loadingText: '加载中',
    }

    const wrapper = render(<Button {...props}>按钮</Button>)

    expect(wrapper.getByText('加载中')).toBeTruthy()
    expect(wrapper.container.innerHTML).toMatchSnapshot()
  })

  it('should render loading of a specific size when using loading-size prop', () => {
    const props = {
      loading: true,
      loadingSize: '10px',
    }

    const wrapper = render(<Button {...props}>按钮</Button>)
    const loading = wrapper.container.querySelector('.rv-loading__spinner')

    expect(loading).toHaveStyle({ width: '10px', height: '10px' })
    expect(wrapper.container.innerHTML).toMatchSnapshot()
  })
})
