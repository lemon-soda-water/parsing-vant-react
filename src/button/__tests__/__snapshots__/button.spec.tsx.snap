// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button test Successful rendering of the button component 1`] = `
<button class="rv-button rv-button--default rv-button--normal"
        type="button"
>
  <div class="rv-button__content">
    <span class="rv-button__text">
      按钮
    </span>
  </div>
</button>
`;

exports[`button test should hide border when color is gradient 1`] = `
<button class="rv-button rv-button--default rv-button--normal"
        style="color: rgb(255, 255, 255); border: 0px;"
        type="button"
>
  <div class="rv-button__content">
    <span class="rv-button__text">
      按钮
    </span>
  </div>
</button>
`;

exports[`button test should not trigger click event when disabled 1`] = `
<button disabled
        class="rv-button rv-button--default rv-button--normal rv-button--disabled"
        type="button"
>
  <div class="rv-button__content">
    <span class="rv-button__text">
      按钮
    </span>
  </div>
</button>
`;

exports[`button test should not trigger click event when loading 1`] = `
<button class="rv-button rv-button--default rv-button--normal rv-button--loading"
        type="button"
>
  <div class="rv-button__content">
    <div class="rv-button__loading rv-loading rv-loading--circular">
      <span class="rv-loading__spinner rv-loading__spinner--circular"
            style="width: 20px; height: 20px;"
      >
        <svg class="rv-loading__circular"
             viewbox="25 25 50 50"
        >
          <circle cx="50"
                  cy="50"
                  r="20"
                  fill="none"
          >
          </circle>
        </svg>
      </span>
    </div>
  </div>
</button>
`;

exports[`button test should render loading of a specific size when using loading-size prop 1`] = `
<button class="rv-button rv-button--default rv-button--normal rv-button--loading"
        type="button"
>
  <div class="rv-button__content">
    <div class="rv-button__loading rv-loading rv-loading--circular">
      <span class="rv-loading__spinner rv-loading__spinner--circular"
            style="width: 10px; height: 10px;"
      >
        <svg class="rv-loading__circular"
             viewbox="25 25 50 50"
        >
          <circle cx="50"
                  cy="50"
                  r="20"
                  fill="none"
          >
          </circle>
        </svg>
      </span>
    </div>
  </div>
</button>
`;

exports[`button test should render loading slot correctly 1`] = `
<button class="rv-button rv-button--default rv-button--normal rv-button--loading"
        type="button"
>
  <div class="rv-button__content">
    <div class="rv-button__loading rv-loading rv-loading--circular">
      <span class="rv-loading__spinner rv-loading__spinner--circular"
            style="width: 20px; height: 20px;"
      >
        <svg class="rv-loading__circular"
             viewbox="25 25 50 50"
        >
          <circle cx="50"
                  cy="50"
                  r="20"
                  fill="none"
          >
          </circle>
        </svg>
      </span>
    </div>
    <span class="rv-button__text">
      加载中
    </span>
  </div>
</button>
`;

exports[`button test should trigger a click event 1`] = `
<button class="rv-button rv-button--default rv-button--normal"
        type="button"
>
  <div class="rv-button__content">
    <span class="rv-button__text">
      按钮
    </span>
  </div>
</button>
`;
